airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -117,70 +229,120 @@ def _evaluate_trigger_rule(
+                select(TaskInstance.task_id, func.count(TaskInstance.task_id))
+                .where(TaskInstance.dag_id == ti.dag_id, TaskInstance.run_id == ti.run_id)
+                .where(
+                    or_(
+                        *self._iter_upstream_conditions(
+                            ti=ti, relevant_tasks=indirect_setups, session=session
+                        )
+                    )
+                )
+                .group_by(TaskInstance.task_id)
+            ).all()
+            upstream = sum(count for _, count in task_id_counts)
 
-            This extra closure allows us to query the database only when needed,
-            and at most once for each task (instead of once for each expanded
-            task instance of the same task).
-            """
-            try:
-                expanded_ti_count = _get_expanded_ti_count()
-            except (NotFullyPopulated, NotMapped):
-                return None
-            return ti.get_relevant_upstream_map_indexes(
-                upstream_tasks[upstream_id],
-                expanded_ti_count,
-                session=session,
-            )
+        new_state = None
+        changed = False
 
-        def _is_relevant_upstream(upstream: TaskInstance) -> bool:
-            """Whether a task instance is a "relevant upstream" of the current task."""
-            # Not actually an upstream task.
-            if upstream.task_id not in task.upstream_task_ids:
-                return False
-            # The current task is not in a mapped task group. All tis from an
-            # upstream task are relevant.
-            if task.get_closest_mapped_task_group() is None:
-                return True
-            # The upstream ti is not expanded. The upstream may be mapped or
-            # not, but the ti is relevant either way.
-            if upstream.map_index < 0:
-                return True
-            # Now we need to perform fine-grained check on whether this specific
-            # upstream ti's map index is relevant.
-            relevant = _get_relevant_upstream_map_indexes(upstream.task_id)
-            if relevant is None:
-                return True
-            if relevant == upstream.map_index:
-                return True
-            if isinstance(relevant, collections.abc.Container) and upstream.map_index in relevant:
-                return True
-            return False
+        # if there's a failure, we mark upstream_failed; if there's a skip, we mark skipped
+        # in either case, we don't wait for all relevant setups to complete
+        if dep_context.flag_upstream_failed:
+            if upstream_failed or failed:
+                new_state = TaskInstanceState.UPSTREAM_FAILED
+            elif skipped:
+                new_state = TaskInstanceState.SKIPPED
+            elif removed and success and ti.map_index > -1:
+                if ti.map_index >= success:
+                    new_state = TaskInstanceState.REMOVED
```

Comment:
> Also IMHO we need a condition on done, otherwise it will not be deterministic.

yeah, the reason why i felt it is ok is that airflow is already a race when there are many-to-one.  so this seems basically consistent with the existing behavior -- to fail or skip fast.  i think in practice it is not likely to make much difference. but...

for others, the scenario is, if there are multiple upstream indirect setups, should we wait for all of them to complete before determining the state of the object task (and then e.g. if there is a combination of success / skip / fail, then we fail it) or should we just take the state of the first non-success?

taking the state of the first non-success is how things work currently when there's a many-to-one direct upstream e.g. `[a, b] >> c`.  in this way it "fails fast" (or skips).  but indirect upstream setup is of course something a little different.  we could choose to wait until all upstream setups are done before marking state.  this means we do *not* fail fast, and the payoff is there is no race condition w.r.t. state.

cc @jedcunningham @ephraimbuddy @potiuk

Reply:
The main problem here is that the skipped state is used as one of the successful states (along with success), most of the users do not consider it as failure, it simply means that there is no need to run the task, nor to send a notification (on_failure).

And the problem will be much complicated if we discuss scenarios where there is an external task sensor waits for this task...

I'd love to hear from others what they think.

Reply:
so with your example, if the upstream setups are skipped, the task will be skipped. it's only if there's a mixed state, which is actually a pretty odd situation, don't think super common generally, but we must to decide what to do.
the question is take the first one or wait for all and the evaluate.
if your position is "wait for all and evaluate" what would be your logic?
```
if at least one failure, then upstream_failed
else if at least one skip, skip
```
?
i think it's probably fine... add an implicit all_done for the setups and then evaluate.  i don't think it makes a _huge_ difference either way.

Reply:
I lean more toward consistency here (fast fail), even thought that consistency is a non-deterministic outcome.